File: RootTools2/src/main/java/com/rarnu/tools/neo/utils/AppUtils.java
package com.rarnu.tools.neo.utils;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Environment;
import com.rarnu.tools.neo.api.NativeAPI;
import com.rarnu.tools.neo.data.AppInfo;

import java.util.ArrayList;
import java.util.List;

public class AppUtils {

    public static boolean isAppRequiredBySystem(String pkgName) {
        return NativeAPI.isAppRequiredBySystem(pkgName);
    }

    public static List<AppInfo> getSystemApps(Context ctx) {
        PackageManager pm = ctx.getPackageManager();
        List<PackageInfo> pkgs = pm.getInstalledPackages(0);

        List<AppInfo> list = new ArrayList<>();
        List<AppInfo> listData = new ArrayList<>();
        List<AppInfo> listDisabled = new ArrayList<>();
        if (pkgs != null) {
            for (PackageInfo pkg : pkgs) {
                if ((pkg.applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) > 0) {
                    if (pkg.applicationInfo.enabled) {
                        list.add(new AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                false,
                                pkg.versionName,
                                pkg.versionCode,
                                true,
                                true
                        ));
                    } else {
                        listDisabled.add(new AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                true,
                                pkg.versionName,
                                pkg.versionCode,
                                true,
                                true
                        ));
                    }
                } else {
                    if (pkg.applicationInfo.enabled) {
                        listData.add(new AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                false,
                                pkg.versionName,
                                pkg.versionCode,
                                false,
                                true
                        ));
                    } else {
                        listDisabled.add(new AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                true,
                                pkg.versionName,
                                pkg.versionCode,
                                false,
                                true
                        ));
                    }
                }
            }
        }
        list.addAll(listData);
        list.addAll(listDisabled);
        return list;
    }

    public static List<AppInfo> getInstalledApps(Context ctx) {
        PackageManager pm = ctx.getPackageManager();
        List<PackageInfo> pkgs = pm.getInstalledPackages(0);
        List<AppInfo> list = new ArrayList<>();
        List<AppInfo> listSystem = new ArrayList<>();
        if (pkgs != null) {
            for (PackageInfo pkg : pkgs) {
                if (pkg.applicationInfo.enabled) {
                    if ((pkg.applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) > 0) {
                        listSystem.add(new AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                true,
                                pkg.versionName,
                                pkg.versionCode,
                                true,
                                false
                        ));
                    } else {
                        list.add(new AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                true,
                                pkg.versionName,
                                pkg.versionCode,
                                false,
                                false
                        ));
                    }

                }
            }
        }
        list.addAll(listSystem);
        return list;
    }

    public static boolean isMIUI(Context ctx) {
        PackageManager pm = ctx.getPackageManager();
        String[] pkgs = new String[]{"com.miui.core", "com.miui.system"};
        boolean isMIUI = true;
        PackageInfo pi;
        for (String s : pkgs) {
            try {
                pi = pm.getPackageInfo(s, 0);
                isMIUI = pi != null;
            } catch (Exception e) {
                isMIUI = false;
            }
            if (!isMIUI) {
                break;
            }
        }
        return isMIUI;
    }

    public static void doScanMedia(Context context) {
        context.sendBroadcast(new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.parse("file://" + Environment.getExternalStorageDirectory().getAbsolutePath())));
    }

}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/utils/AppUtils.kt
package com.rarnu.tools.neo.utils

import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.net.Uri
import android.os.Environment
import com.rarnu.tools.neo.data.AppInfo

object AppUtils {

    fun getSystemApps(ctx: Context?): MutableList<AppInfo> {
        val pm = ctx?.packageManager
        val pkgs = pm?.getInstalledPackages(0)
        val list = arrayListOf<AppInfo>()
        val listData = arrayListOf<AppInfo>()
        val listDisabled = arrayListOf<AppInfo>()
        if (pkgs != null) {
            for (pkg in pkgs) {
                if (pkg.applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM > 0) {
                    if (pkg.applicationInfo.enabled) {
                        list.add(AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                false,
                                pkg.versionName,
                                pkg.versionCode,
                                true,
                                true))
                    } else {
                        listDisabled.add(AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                true,
                                pkg.versionName,
                                pkg.versionCode,
                                true,
                                true))
                    }
                } else {
                    if (pkg.applicationInfo.enabled) {
                        listData.add(AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                false,
                                pkg.versionName,
                                pkg.versionCode,
                                false,
                                true))
                    } else {
                        listDisabled.add(AppInfo(
                                pkg.applicationInfo.loadLabel(pm).toString(),
                                pkg.applicationInfo.loadIcon(pm),
                                pkg.packageName,
                                true,
                                pkg.versionName,
                                pkg.versionCode,
                                false,
                                true))
                    }
                }
            }
        }
        list.sortBy { it.name }
        if (listData.size > 0) {
            listData.sortBy { it.name }
            list.addAll(listData)
        }
        if (listDisabled.size > 0) {
            listDisabled.sortBy { it.name }
            list.addAll(listDisabled)
        }
        return list
    }

    fun getInstalledApps(ctx: Context?): MutableList<AppInfo> {
        val pm = ctx?.packageManager
        val pkgs = pm?.getInstalledPackages(0)
        val list = arrayListOf<AppInfo>()
        val listSystem = arrayListOf<AppInfo>()
        if (pkgs != null) {
            pkgs.filter { it.applicationInfo.enabled }.forEach {
                if (it.applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM > 0) {
                    listSystem.add(AppInfo(
                            it.applicationInfo.loadLabel(pm).toString(),
                            it.applicationInfo.loadIcon(pm),
                            it.packageName,
                            true,
                            it.versionName,
                            it.versionCode,
                            true,
                            false))
                } else {
                    list.add(AppInfo(
                            it.applicationInfo.loadLabel(pm).toString(),
                            it.applicationInfo.loadIcon(pm),
                            it.packageName,
                            true,
                            it.versionName,
                            it.versionCode,
                            false,
                            false))
                }
            }
        }
        list.sortBy { it.name }
        if (listSystem.size > 0) {
            listSystem.sortBy { it.name }
            list.addAll(listSystem)
        }
        return list
    }

    fun isMIUI(ctx: Context?): Boolean {
        val pm = ctx?.packageManager
        val pkgs = arrayOf("com.miui.core", "com.miui.system")
        var isMIUI = true
        var pi: PackageInfo?
        for (s in pkgs) {
            try {
                pi = pm?.getPackageInfo(s, 0)
                isMIUI = pi != null
            } catch (e: Exception) {
                isMIUI = false
            }

            if (!isMIUI) {
                break
            }
        }
        return isMIUI
    }

    fun doScanMedia(context: Context?) =
            context?.sendBroadcast(Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.parse("file://" + Environment.getExternalStorageDirectory().absolutePath)))

}
